scripts/gen/tpcc.sh

```diff
@@ -118,33 +171,154 @@ then
   echo "done importing"
 fi
 
-# if f_inc is not specified, then we set it equal to f_warehouses, this means
-# we just need to run one iteration.
-if [[ $f_inc == 0 ]];
+# Comment out since we don't have a scenario using this at present.
+## if f_inc is not specified, then we set it equal to f_warehouses, this means
+## we just need to run one iteration.
+#if [[ $f_inc == 0 ]];
+#then
+#  f_inc=$f_warehouses
+#fi
+
+# We first run an exploration test round with binary search to find a right
+# range of test boundary to use in the subsequent official run.
+# For certain machine types the binary search result don't work well and will
+# cause kv errors and end the test, so we need to handle them differently.
+tpcc_files=""
+has_good_result=0
+if [[ special_type -eq 0 ]];
 then
-  f_inc=$f_warehouses
+  l=$f_active
+  r=$f_warehouses
+  while [[ $l -le $r ]]
+  do
+      mid=$((l + ((r - l)/2)))
+      report="${logdir}/tpcc-results-$mid.txt"
+      ./cockroach workload run tpcc \
+          --warehouses="$mid"  \
+          --active-warehouses="$mid" \
+          --ramp=1m --duration="1m" \
+          "${pgurls[@]}" > "$report"
+
+      # awk uses lexicographic order by default when doing comparison, we
+      # need to force it to use number comparison.
+      if [[ $(tail -1 "$report" | awk '{if($3+0 > 87){print "pass"}}') != "pass" ]];
+      then
+        echo "Test TPCC on $mid failed"
+        ((r=mid-f_inc))
+        current_result_good=0
+      else
+        echo "Test TPCC on $mid passed"
+        ((l=mid+f_inc))
+        current_result_good=1
+        if [[ $mid -gt best_result ]]
+        then
+          best_result=$mid
+        fi
+      fi
+  done
 fi
 
-for active in `seq $f_active $f_inc $f_warehouses`
-do
-  echo "Running TPCC: $active"
-  report="${logdir}/tpcc-results-$active.txt"
-./cockroach workload run tpcc \
-    --warehouses="$f_warehouses"  \
-    --active-warehouses="$active" \
-    --ramp=1m --duration="$f_duration" \
-    "${pgurls[@]}" > "$report"
-
-    if [[ $(tail -1 "$report" | awk '{if($3 > 85 && $7 < 10000){print "pass"}}') != "pass" ]];
+result_files=(`find "${logdir}"/ -maxdepth 1 -name "tpcc-results*.txt"`)
+if [ ${#result_files[@]} -gt 0 ];
+then
+  rm "${logdir}"/tpcc-results*.txt
+  echo "Deleted exploration run result."
+fi
+
+
+# if valid result is found in the previous test round, we start an official
+# run after adjusting of the starting point to a multiple of one hundred.
+# Because at present tpcc is pretty unstable where result can fluctuate
+# within quite a range, we retry a few times with lower boundary for the
+# official run stage before we call the test fail.
+#
+# In order to achieve better comparison result across multiple machine
+# types, we use a small increase of store size of 50.
+if [ $best_result -gt 0 ]
+then
+  ((l=$best_result-$best_result%f_inc))
+  if [ $current_result_good -eq 0 ]
+  then
+    ((l-=f_inc))
+  fi
+  ((l-=f_inc*7))
+  ((h=l+100))
```

Comment:
What's the rationale behind the magic multiplier `7` and setting `h` below `best_result`?

Reply:
This is because of short run result doesn't reflect the long result run, there was a gap between the two, this difference reflects that.

Reply:
`f_inc` is user-specified; shouldn't the multiplier be scaled? otherwise, for a large value of `f_inc`, the multiplier could overshoot.